framework/src/main/java/org/checkerframework/framework/util/defaults/QualifierDefaults.java

```diff
@@ -629,55 +700,39 @@ private DefaultSet defaultsAt(final Element elt) {
             return elementDefaults.get(elt);
         }
 
-        DefaultSet qualifiers = null;
-
-        {
-            AnnotationMirror dqAnno = atypeFactory.getDeclAnnotation(elt, DefaultQualifier.class);
-
-            if (dqAnno != null) {
-                qualifiers = new DefaultSet();
-                Set<Default> p = fromDefaultQualifier(dqAnno);
-
-                if (p != null) {
-                    qualifiers.addAll(p);
-                }
-            }
-        }
+        DefaultSet qualifiers = defaultsAtDirect(elt);
 
-        {
-            AnnotationMirror dqListAnno =
-                    atypeFactory.getDeclAnnotation(elt, DefaultQualifier.List.class);
-            if (dqListAnno != null) {
-                if (qualifiers == null) {
-                    qualifiers = new DefaultSet();
-                }
-                @SuppressWarnings("unchecked")
-                List<AnnotationMirror> values =
-                        AnnotationUtils.getElementValue(
-                                dqListAnno, defaultQualifierListValueElement, List.class);
-                for (AnnotationMirror dqAnno : values) {
-                    Set<Default> p = fromDefaultQualifier(dqAnno);
-                    if (p != null) {
-                        qualifiers.addAll(p);
-                    }
+        DefaultSet parentDefaults;
+        if (elt.getKind() == ElementKind.PACKAGE) {
+            Element parent = ElementUtils.parentPackage((PackageElement) elt, elements);
+            DefaultSet origParentDefaults = defaultsAt(parent);
+            parentDefaults = new DefaultSet();
+            for (Default d : origParentDefaults) {
+                if (d.applyToSubpackages) {
+                    parentDefaults.add(d);
                 }
             }
-        }
-
-        Element parent;
-        if (elt.getKind() == ElementKind.PACKAGE) {
-            parent = ElementUtils.parentPackage((PackageElement) elt, elements);
         } else {
-            parent = elt.getEnclosingElement();
+            Element parent = elt.getEnclosingElement();
+            parentDefaults = defaultsAt(parent);
         }
 
-        DefaultSet parentDefaults = defaultsAt(parent);
         if (qualifiers == null || qualifiers.isEmpty()) {
             qualifiers = parentDefaults;
         } else {
+            // TODO(cpovirk): What should happen with conflicts?
             qualifiers.addAll(parentDefaults);
         }
 
+        /* TODO: it would seem more efficient to also cache null/empty as the result.
```

Comment:
I was wondering this, too :) I wondered if maybe other code was modifying the set somewhere (which would result in modifying `EMPTY`). I also wondered if we'd just end up with a ton of entries in the map, potentially using a lot of memory. Granted, if caching empty defaults for a lot of code would cause a memory problem now, then caching `@NullMarked` defaults for a lot of code may cause that same problem later :)

(If so, part of the answer may be a singleton `DefaultSet.NULL_MARKED`. But even with that, I'd expect an entry in `elementDefaults` for almost every `Element`.....)